<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="prev" title="singler" href="modules.html" />

    <!-- Generated with Sphinx 7.2.6 and Furo 2023.09.10 -->
        <title>singler package - singler unknown documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">singler unknown documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">singler unknown documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="modules.html">Module Reference</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Module Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">singler package</a></li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="singler-package">
<h1>singler package<a class="headerlink" href="#singler-package" title="Link to this heading">#</a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading">#</a></h2>
</section>
<section id="module-singler.annotate_integrated">
<span id="singler-annotate-integrated-module"></span><h2>singler.annotate_integrated module<a class="headerlink" href="#module-singler.annotate_integrated" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="singler.annotate_integrated.annotate_integrated">
<span class="sig-prename descclassname"><span class="pre">singler.annotate_integrated.</span></span><span class="sig-name descname"><span class="pre">annotate_integrated</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">test_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_data_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_labels_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_features_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_check_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_check_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">build_single_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classify_single_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">build_integrated_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classify_integrated_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/annotate_integrated.html#annotate_integrated"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.annotate_integrated.annotate_integrated" title="Link to this definition">#</a></dt>
<dd><p>Annotate a single-cell expression dataset based on the correlation
of each cell to profiles in multiple labelled references, where the
annotation from each reference is then integrated across references.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>test_data</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>) – <p>A matrix-like object representing the test dataset, where rows are
features and columns are samples (usually cells). Entries should be expression
values; only the ranking within each column will be used.</p>
<p>Alternatively, a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
containing such a matrix in one of its assays.</p>
</p></li>
<li><p><strong>test_features</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>) – Sequence of length equal to the number of rows in
<code class="docutils literal notranslate"><span class="pre">test_data</span></code>, containing the feature identifier for each row.</p></li>
<li><p><strong>ref_data_list</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]) – <p>Sequence consisting of one or more of the following:</p>
<ul>
<li><p>A matrix-like object representing the reference dataset, where rows
are features and columns are samples. Entries should be expression values,
usually log-transformed (see comments for the <code class="docutils literal notranslate"><span class="pre">ref</span></code> argument in
<a class="reference internal" href="#singler.build_single_reference.build_single_reference" title="singler.build_single_reference.build_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a>).</p></li>
<li><p>A
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
object containing such a matrix in its assays.</p></li>
<li><p>A string that can be passed as <code class="docutils literal notranslate"><span class="pre">name</span></code> to
<a class="reference internal" href="#singler.fetch_reference.fetch_github_reference" title="singler.fetch_reference.fetch_github_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">fetch_github_reference()</span></code></a>.
This will use the specified dataset as the reference.</p></li>
</ul>
</p></li>
<li><p><strong>ref_labels_list</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]]) – <p>Sequence of the same length as <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>, where the contents
depend on the type of value in the corresponding entry of <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>:</p>
<ul>
<li><p>If <code class="docutils literal notranslate"><span class="pre">ref_data_list[i]</span></code> is a matrix-like object, <code class="docutils literal notranslate"><span class="pre">ref_labels_list[i]</span></code> should be
a sequence of length equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">ref_data_list[i]</span></code>,
containing the label associated with each column.</p></li>
<li><p>If <code class="docutils literal notranslate"><span class="pre">ref_data_list[i]</span></code> is a string, <code class="docutils literal notranslate"><span class="pre">ref_labels_list[i]</span></code> should be a string
specifying the label type to use, e.g., “main”, “fine”, “ont”.
If a single string is supplied, it is recycled for all <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>.</p></li>
</ul>
</p></li>
<li><p><strong>ref_features_list</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]]) – <p>Sequence of the same length as <code class="docutils literal notranslate"><span class="pre">ref_data_list</span></code>, where the contents
depend on the type of value in the corresponding entry of <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>:</p>
<ul>
<li><p>If <code class="docutils literal notranslate"><span class="pre">ref_data_list[i]</span></code> is a matrix-like object, <code class="docutils literal notranslate"><span class="pre">ref_features_list[i]</span></code> should be
a sequence of length equal to the number of rows of <code class="docutils literal notranslate"><span class="pre">ref_data_list[i]</span></code>,
containing the feature identifier associated with each row.</p></li>
<li><p>If <code class="docutils literal notranslate"><span class="pre">ref_data_list[i]</span></code> is a string, <code class="docutils literal notranslate"><span class="pre">ref_features_list[i]</span></code> should be a string
specifying the feature type to use, e.g., “ensembl”, “symbol”.
If a single string is supplied, it is recycled for all <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>.</p></li>
</ul>
</p></li>
<li><p><strong>test_assay_type</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]) – Assay of <code class="docutils literal notranslate"><span class="pre">test_data</span></code> containing the expression matrix, if <code class="docutils literal notranslate"><span class="pre">test_data</span></code> is a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p></li>
<li><p><strong>test_check_missing</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Whether to check for and remove missing (i.e., NaN) values from the test dataset.</p></li>
<li><p><strong>ref_assay_type</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]) – Assay containing the expression matrix for any entry of <code class="docutils literal notranslate"><span class="pre">ref_data_list</span></code> that is a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p></li>
<li><p><strong>ref_check_missing</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Whether to check for and remove missing (i.e., NaN) values from the reference datasets.</p></li>
<li><p><strong>cache_dir</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – Path to a cache directory for downloading reference files, see
<a class="reference internal" href="#singler.fetch_reference.fetch_github_reference" title="singler.fetch_reference.fetch_github_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">fetch_github_reference()</span></code></a> for details.
Only used if <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> is a string.</p></li>
<li><p><strong>build_single_args</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>) – Further arguments to pass to
<a class="reference internal" href="#singler.build_single_reference.build_single_reference" title="singler.build_single_reference.build_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a>.</p></li>
<li><p><strong>classify_single_args</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>) – Further arguments to pass to
<a class="reference internal" href="#singler.classify_single_reference.classify_single_reference" title="singler.classify_single_reference.classify_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">classify_single_reference()</span></code></a>.</p></li>
<li><p><strong>build_integrated_args</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>) – Further arguments to pass to
<a class="reference internal" href="#singler.build_integrated_references.build_integrated_references" title="singler.build_integrated_references.build_integrated_references"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_integrated_references()</span></code></a>.</p></li>
<li><p><strong>classify_integrated_args</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>) – Further arguments to pass to
<a class="reference internal" href="#singler.classify_integrated_references.classify_integrated_references" title="singler.classify_integrated_references.classify_integrated_references"><code class="xref py py-meth docutils literal notranslate"><span class="pre">classify_integrated_references()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Number of threads to use for the various steps.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Tuple" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#list" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>], <a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Tuple where the first element contains per-reference results (i.e. a
list of BiocFrame outputs equivalent to running
<a class="reference internal" href="#singler.annotate_single.annotate_single" title="singler.annotate_single.annotate_single"><code class="xref py py-meth docutils literal notranslate"><span class="pre">annotate_single()</span></code></a> on each reference)
and the second element contains integrated results across references
(i.e., a BiocFrame from
<a class="reference internal" href="#singler.classify_integrated_references.classify_integrated_references" title="singler.classify_integrated_references.classify_integrated_references"><code class="xref py py-meth docutils literal notranslate"><span class="pre">classify_integrated_references()</span></code></a>).</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler.annotate_single">
<span id="singler-annotate-single-module"></span><h2>singler.annotate_single module<a class="headerlink" href="#module-singler.annotate_single" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="singler.annotate_single.annotate_single">
<span class="sig-prename descclassname"><span class="pre">singler.annotate_single.</span></span><span class="sig-name descname"><span class="pre">annotate_single</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">test_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">build_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classify_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/annotate_single.html#annotate_single"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.annotate_single.annotate_single" title="Link to this definition">#</a></dt>
<dd><p>Annotate a single-cell expression dataset based on the correlation
of each cell to profiles in a labelled reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>test_data</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>) – <p>A matrix-like object representing the test dataset, where rows are
features and columns are samples (usually cells). Entries should be expression
values; only the ranking within each column will be used.</p>
<p>Alternatively, a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
containing such a matrix in one of its assays. Non-default assay
types can be specified in <code class="docutils literal notranslate"><span class="pre">classify_args</span></code>.</p>
</p></li>
<li><p><strong>test_features</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>) – Sequence of length equal to the number of rows in
<code class="docutils literal notranslate"><span class="pre">test_data</span></code>, containing the feature identifier for each row.</p></li>
<li><p><strong>ref_data</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>) – <p>A matrix-like object representing the reference dataset, where rows
are features and columns are samples. Entries should be expression values,
usually log-transformed (see comments for the <code class="docutils literal notranslate"><span class="pre">ref</span></code> argument in
<a class="reference internal" href="#singler.build_single_reference.build_single_reference" title="singler.build_single_reference.build_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a>).</p>
<p>Alternatively, a string that can be passed as <code class="docutils literal notranslate"><span class="pre">name</span></code> to
<a class="reference internal" href="#singler.fetch_reference.fetch_github_reference" title="singler.fetch_reference.fetch_github_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">fetch_github_reference()</span></code></a>.
This will use the specified dataset as the reference.</p>
</p></li>
<li><p><strong>ref_labels</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – <p>If <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> is a matrix-like object, <code class="docutils literal notranslate"><span class="pre">ref_labels</span></code> should be
a sequence of length equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>,
containing the label associated with each column.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> is a string, <code class="docutils literal notranslate"><span class="pre">ref_labels</span></code> should be a string
specifying the label type to use, e.g., “main”, “fine”, “ont”.</p>
</p></li>
<li><p><strong>ref_features</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – <p>If <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> is a matrix-like object, <code class="docutils literal notranslate"><span class="pre">ref_features</span></code> should be
a sequence of length equal to the number of rows of <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>,
containing the feature identifier associated with each row.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> is a string, <code class="docutils literal notranslate"><span class="pre">ref_features</span></code> should be a string
specifying the label type to use, e.g., “ensembl”, “symbol”.</p>
</p></li>
<li><p><strong>cache_dir</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – Path to a cache directory for downloading reference files, see
<a class="reference internal" href="#singler.fetch_reference.fetch_github_reference" title="singler.fetch_reference.fetch_github_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">fetch_github_reference()</span></code></a> for details.
Only used if <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> is a string.</p></li>
<li><p><strong>build_args</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>) – Further arguments to pass to
<a class="reference internal" href="#singler.build_single_reference.build_single_reference" title="singler.build_single_reference.build_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a>.</p></li>
<li><p><strong>classify_args</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>) – Further arguments to pass to
<a class="reference internal" href="#singler.classify_single_reference.classify_single_reference" title="singler.classify_single_reference.classify_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">classify_single_reference()</span></code></a>.</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Number of threads to use for the various steps.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A data frame containing the labelling results, see
<a class="reference internal" href="#singler.classify_single_reference.classify_single_reference" title="singler.classify_single_reference.classify_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">classify_single_reference()</span></code></a>
for details. The metadata also contains a <code class="docutils literal notranslate"><span class="pre">markers</span></code> dictionary,
specifying the markers that were used for each pairwise comparison
between labels; and a list of <code class="docutils literal notranslate"><span class="pre">unique_markers</span></code> across all labels.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler.build_integrated_references">
<span id="singler-build-integrated-references-module"></span><h2>singler.build_integrated_references module<a class="headerlink" href="#module-singler.build_integrated_references" title="Link to this heading">#</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="singler.build_integrated_references.IntegratedReferences">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">singler.build_integrated_references.</span></span><span class="sig-name descname"><span class="pre">IntegratedReferences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_features</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/build_integrated_references.html#IntegratedReferences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.build_integrated_references.IntegratedReferences" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#object" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Object containing integrated references, typically constructed by
<a class="reference internal" href="#singler.build_integrated_references.build_integrated_references" title="singler.build_integrated_references.build_integrated_references"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_integrated_references()</span></code></a>.</p>
<dl class="py property">
<dt class="sig sig-object py" id="singler.build_integrated_references.IntegratedReferences.reference_labels">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference_labels</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#list" title="(in Python v3.9)"><span class="pre">list</span></a></em><a class="headerlink" href="#singler.build_integrated_references.IntegratedReferences.reference_labels" title="Link to this definition">#</a></dt>
<dd><p>List of lists containing the names of the labels for each reference.</p>
<p>Each entry corresponds to a reference in <a class="reference internal" href="#singler.build_integrated_references.IntegratedReferences.reference_names" title="singler.build_integrated_references.IntegratedReferences.reference_names"><code class="xref py py-attr docutils literal notranslate"><span class="pre">reference_names</span></code></a>,
if <code class="docutils literal notranslate"><span class="pre">reference_names</span></code> is not None.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="singler.build_integrated_references.IntegratedReferences.reference_names">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">reference_names</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><span class="pre">Sequence</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/constants.html#None" title="(in Python v3.9)"><span class="pre">None</span></a></em><a class="headerlink" href="#singler.build_integrated_references.IntegratedReferences.reference_names" title="Link to this definition">#</a></dt>
<dd><p>Sequence containing the names of the references. Alternatively
None, if no names were supplied.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="singler.build_integrated_references.IntegratedReferences.test_features">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">test_features</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#list" title="(in Python v3.9)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#singler.build_integrated_references.IntegratedReferences.test_features" title="Link to this definition">#</a></dt>
<dd><p>Sequence containing the names of the test features.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="singler.build_integrated_references.build_integrated_references">
<span class="sig-prename descclassname"><span class="pre">singler.build_integrated_references.</span></span><span class="sig-name descname"><span class="pre">build_integrated_references</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">test_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_data_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_labels_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_features_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_prebuilt_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/build_integrated_references.html#build_integrated_references"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.build_integrated_references.build_integrated_references" title="Link to this definition">#</a></dt>
<dd><p>Build a set of integrated references for classification of a test dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>test_features</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>) – Sequence of features for the test dataset.</p></li>
<li><p><strong>ref_data_list</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>) – List of reference datasets, where each entry is equivalent to <code class="docutils literal notranslate"><span class="pre">ref_data</span></code> in
<a class="reference internal" href="#singler.build_single_reference.build_single_reference" title="singler.build_single_reference.build_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a>.</p></li>
<li><p><strong>ref_labels_list</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#list" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]) – List of reference labels, where each entry is equivalent to <code class="docutils literal notranslate"><span class="pre">ref_labels</span></code> in
<a class="reference internal" href="#singler.build_single_reference.build_single_reference" title="singler.build_single_reference.build_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a>.</p></li>
<li><p><strong>ref_features_list</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#list" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]) – List of reference features, where each entry is equivalent to <code class="docutils literal notranslate"><span class="pre">ref_features</span></code> in
<a class="reference internal" href="#singler.build_single_reference.build_single_reference" title="singler.build_single_reference.build_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a>.</p></li>
<li><p><strong>ref_prebuilt_list</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#list" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference internal" href="#singler.build_single_reference.SinglePrebuiltReference" title="singler.build_single_reference.SinglePrebuiltReference"><code class="xref py py-class docutils literal notranslate"><span class="pre">SinglePrebuiltReference</span></code></a>]) – List of prebuilt references, typically created by
calling <a class="reference internal" href="#singler.build_single_reference.build_single_reference" title="singler.build_single_reference.build_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a> on the corresponding
elements of <code class="docutils literal notranslate"><span class="pre">ref_data_list</span></code>, <code class="docutils literal notranslate"><span class="pre">ref_labels_list</span></code> and <code class="docutils literal notranslate"><span class="pre">ref_features_list</span></code>.</p></li>
<li><p><strong>ref_names</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]]) – Sequence of names for the references.
If None, these are automatically generated.</p></li>
<li><p><strong>assay_type</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]) – Assay containing the expression matrix for any entry of <code class="docutils literal notranslate"><span class="pre">ref_data_list</span></code> that is a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p></li>
<li><p><strong>check_missing</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Whether to check for and remove rows with missing (NaN) values
from each entry of <code class="docutils literal notranslate"><span class="pre">ref_data_list</span></code>.</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Number of threads.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#singler.build_integrated_references.IntegratedReferences" title="singler.build_integrated_references.IntegratedReferences"><code class="xref py py-class docutils literal notranslate"><span class="pre">IntegratedReferences</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Integrated references for classification with
<a class="reference internal" href="#singler.classify_integrated_references.classify_integrated_references" title="singler.classify_integrated_references.classify_integrated_references"><code class="xref py py-meth docutils literal notranslate"><span class="pre">classify_integrated_references()</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler.build_single_reference">
<span id="singler-build-single-reference-module"></span><h2>singler.build_single_reference module<a class="headerlink" href="#module-singler.build_single_reference" title="Link to this heading">#</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="singler.build_single_reference.SinglePrebuiltReference">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">singler.build_single_reference.</span></span><span class="sig-name descname"><span class="pre">SinglePrebuiltReference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">markers</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/build_single_reference.html#SinglePrebuiltReference"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.build_single_reference.SinglePrebuiltReference" title="Link to this definition">#</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#object" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>A prebuilt reference object, typically created by
<a class="reference internal" href="#singler.build_single_reference.build_single_reference" title="singler.build_single_reference.build_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a>. This is intended for advanced users only and
should not be serialized.</p>
<dl class="py property">
<dt class="sig sig-object py" id="singler.build_single_reference.SinglePrebuiltReference.features">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">features</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><span class="pre">Sequence</span></a></em><a class="headerlink" href="#singler.build_single_reference.SinglePrebuiltReference.features" title="Link to this definition">#</a></dt>
<dd><p>Returns:
The universe of features known to this reference, usually as strings.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="singler.build_single_reference.SinglePrebuiltReference.labels">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">labels</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><span class="pre">Sequence</span></a></em><a class="headerlink" href="#singler.build_single_reference.SinglePrebuiltReference.labels" title="Link to this definition">#</a></dt>
<dd><p>Returns:
Unique labels in this reference.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="singler.build_single_reference.SinglePrebuiltReference.marker_subset">
<span class="sig-name descname"><span class="pre">marker_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">indices_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/build_single_reference.html#SinglePrebuiltReference.marker_subset"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.build_single_reference.SinglePrebuiltReference.marker_subset" title="Link to this definition">#</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>indices_only</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Whether to return the markers as indices
into <a class="reference internal" href="#singler.build_single_reference.SinglePrebuiltReference.features" title="singler.build_single_reference.SinglePrebuiltReference.features"><code class="xref py py-attr docutils literal notranslate"><span class="pre">features</span></code></a>, or as a list of feature identifiers.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ndarray</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#list" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>If <code class="docutils literal notranslate"><span class="pre">indices_only</span> <span class="pre">=</span> <span class="pre">False</span></code>, a list of feature identifiers for the markers.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">indices_only</span> <span class="pre">=</span> <span class="pre">True</span></code>, a NumPy array containing the integer indices of
features in <code class="docutils literal notranslate"><span class="pre">features</span></code> that were chosen as markers.</p>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="singler.build_single_reference.SinglePrebuiltReference.markers">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">markers</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><span class="pre">dict</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><span class="pre">Any</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><span class="pre">dict</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><span class="pre">Any</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><span class="pre">Sequence</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#singler.build_single_reference.SinglePrebuiltReference.markers" title="Link to this definition">#</a></dt>
<dd><p>Returns:
Markers for every pairwise comparison between labels.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="singler.build_single_reference.SinglePrebuiltReference.num_labels">
<span class="sig-name descname"><span class="pre">num_labels</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/build_single_reference.html#SinglePrebuiltReference.num_labels"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.build_single_reference.SinglePrebuiltReference.num_labels" title="Link to this definition">#</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Number of unique labels in this reference.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="singler.build_single_reference.SinglePrebuiltReference.num_markers">
<span class="sig-name descname"><span class="pre">num_markers</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/build_single_reference.html#SinglePrebuiltReference.num_markers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.build_single_reference.SinglePrebuiltReference.num_markers" title="Link to this definition">#</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Number of markers to be used for classification. This is the
same as the size of the array from <a class="reference internal" href="#singler.build_single_reference.SinglePrebuiltReference.marker_subset" title="singler.build_single_reference.SinglePrebuiltReference.marker_subset"><code class="xref py py-meth docutils literal notranslate"><span class="pre">marker_subset()</span></code></a>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="singler.build_single_reference.build_single_reference">
<span class="sig-prename descclassname"><span class="pre">singler.build_single_reference.</span></span><span class="sig-name descname"><span class="pre">build_single_reference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">restrict_to</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">markers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'classic'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_args</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">approximate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/build_single_reference.html#build_single_reference"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.build_single_reference.build_single_reference" title="Link to this definition">#</a></dt>
<dd><p>Build a single reference dataset in preparation for classification.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_data</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>) – <p>A matrix-like object where rows are features, columns are
reference profiles, and each entry is the expression value.
If <cite>markers</cite> is not provided, expression should be normalized
and log-transformed in preparation for marker prioritization via
differential expression analyses. Otherwise, any expression values
are acceptable as only the ranking within each column is used.</p>
<p>Alternatively, a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
containing such a matrix in one of its assays.</p>
</p></li>
<li><p><strong>labels</strong> – Sequence of labels for each reference profile,
i.e., column in <code class="docutils literal notranslate"><span class="pre">ref</span></code>.</p></li>
<li><p><strong>features</strong> – Sequence of identifiers for each feature,
i.e., row in <code class="docutils literal notranslate"><span class="pre">ref</span></code>.</p></li>
<li><p><strong>assay_type</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]) – Assay containing the expression matrix,
if <cite>ref_data</cite> is a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p></li>
<li><p><strong>check_missing</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Whether to check for and remove rows with missing (NaN) values
from <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>.</p></li>
<li><p><strong>restrict_to</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#set" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">set</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/constants.html#None" title="(in Python v3.9)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]) – Subset of available features to restrict to. Only features in
<code class="docutils literal notranslate"><span class="pre">restrict_to</span></code> will be used in the reference building. If None,
no restriction is performed.</p></li>
<li><p><strong>markers</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]]]) – Upregulated markers for each pairwise comparison between labels.
Specifically, <code class="docutils literal notranslate"><span class="pre">markers[a][b]</span></code> should be a sequence of features
that are upregulated in <code class="docutils literal notranslate"><span class="pre">a</span></code> compared to <code class="docutils literal notranslate"><span class="pre">b</span></code>. All such features
should be present in <code class="docutils literal notranslate"><span class="pre">features</span></code>, and all labels in <code class="docutils literal notranslate"><span class="pre">labels</span></code>
should have keys in the inner and outer dictionaries.</p></li>
<li><p><strong>marker_method</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Literal" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'classic'</span></code>]) – Method to identify markers from each pairwise comparisons between
labels in <code class="docutils literal notranslate"><span class="pre">ref_data</span></code>.  If “classic”, we call
<a class="reference internal" href="#singler.get_classic_markers.get_classic_markers" title="singler.get_classic_markers.get_classic_markers"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_classic_markers()</span></code></a>.
Only used if <code class="docutils literal notranslate"><span class="pre">markers</span></code> is not supplied.</p></li>
<li><p><strong>marker_args</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>) – Further arguments to pass to the chosen marker detection method.
Only used if <code class="docutils literal notranslate"><span class="pre">markers</span></code> is not supplied.</p></li>
<li><p><strong>approximate</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Whether to use an approximate neighbor search to compute scores
during classification.</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Number of threads to use for reference building.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#singler.build_single_reference.SinglePrebuiltReference" title="singler.build_single_reference.SinglePrebuiltReference"><code class="xref py py-class docutils literal notranslate"><span class="pre">SinglePrebuiltReference</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The pre-built reference, ready for use in downstream methods like
<a class="reference internal" href="#singler.classify_single_reference.classify_single_reference" title="singler.classify_single_reference.classify_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">classify_single_reference()</span></code></a>.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler.classify_integrated_references">
<span id="singler-classify-integrated-references-module"></span><h2>singler.classify_integrated_references module<a class="headerlink" href="#module-singler.classify_integrated_references" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="singler.classify_integrated_references.classify_integrated_references">
<span class="sig-prename descclassname"><span class="pre">singler.classify_integrated_references.</span></span><span class="sig-name descname"><span class="pre">classify_integrated_references</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">test_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">results</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">integrated_prebuilt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/classify_integrated_references.html#classify_integrated_references"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.classify_integrated_references.classify_integrated_references" title="Link to this definition">#</a></dt>
<dd><p>Integrate classification results across multiple references for a single test dataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>test_data</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>) – <p>A matrix-like object where each row is a feature and each column
is a test sample (usually a single cell), containing expression values.
Normalized and/or transformed expression values are also acceptable as only
the ranking is used within this function.</p>
<p>Alternatively, a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
containing such a matrix in one of its assays.</p>
</p></li>
<li><p><strong>results</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#list" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]]) – List of classification results generated by running
<a class="reference internal" href="#singler.classify_single_reference.classify_single_reference" title="singler.classify_single_reference.classify_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">classify_single_reference()</span></code></a>
on <code class="docutils literal notranslate"><span class="pre">test_data</span></code> with each reference. This may be either the full
data frame or just the <code class="docutils literal notranslate"><span class="pre">&quot;best&quot;</span></code> column. References should be ordered
as in <code class="docutils literal notranslate"><span class="pre">integrated_prebuilt.reference_names</span></code>.</p></li>
<li><p><strong>integrated_prebuilt</strong> (<a class="reference internal" href="#singler.build_integrated_references.IntegratedReferences" title="singler.build_integrated_references.IntegratedReferences"><code class="xref py py-class docutils literal notranslate"><span class="pre">IntegratedReferences</span></code></a>) – Integrated reference object, constructed with
<a class="reference internal" href="#singler.build_integrated_references.build_integrated_references" title="singler.build_integrated_references.build_integrated_references"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_integrated_references()</span></code></a>.</p></li>
<li><p><strong>assay_type</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]) – Assay containing the expression matrix,
if <cite>test_data</cite> is a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p></li>
<li><p><strong>quantile</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#float" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>) – Quantile of the correlation distribution for computing the score for each label.
Larger values increase sensitivity of matches at the expense of
similarity to the average behavior of each label.</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Number of threads to use during classification.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A data frame containing the <code class="docutils literal notranslate"><span class="pre">best_label</span></code> across all
references, defined as the assigned label in the best reference; the
identity of the <code class="docutils literal notranslate"><span class="pre">best_reference</span></code>, either as a name string or an
integer index; the <code class="docutils literal notranslate"><span class="pre">scores</span></code> for each reference, as a nested
BiocFrame; and the <code class="docutils literal notranslate"><span class="pre">delta</span></code> from the best to the second-best
reference. Each row corresponds to a column of <code class="docutils literal notranslate"><span class="pre">test</span></code>.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler.classify_single_reference">
<span id="singler-classify-single-reference-module"></span><h2>singler.classify_single_reference module<a class="headerlink" href="#module-singler.classify_single_reference" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="singler.classify_single_reference.classify_single_reference">
<span class="sig-prename descclassname"><span class="pre">singler.classify_single_reference.</span></span><span class="sig-name descname"><span class="pre">classify_single_reference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">test_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_prebuilt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quantile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_fine_tune</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fine_tune_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/classify_single_reference.html#classify_single_reference"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.classify_single_reference.classify_single_reference" title="Link to this definition">#</a></dt>
<dd><p>Classify a test dataset against a reference by assigning labels from the latter to each column of the former
using the SingleR algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>test_data</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>) – <p>A matrix-like object where each row is a feature and each column
is a test sample (usually a single cell), containing expression values.
Normalized and transformed expression values are also acceptable as only
the ranking is used within this function.</p>
<p>Alternatively, a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
containing such a matrix in one of its assays.</p>
</p></li>
<li><p><strong>test_features</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>) – Sequence of identifiers for each feature in the test
dataset, i.e., row in <code class="docutils literal notranslate"><span class="pre">test_data</span></code>.</p></li>
<li><p><strong>ref_prebuilt</strong> (<a class="reference internal" href="#singler.build_single_reference.SinglePrebuiltReference" title="singler.build_single_reference.SinglePrebuiltReference"><code class="xref py py-class docutils literal notranslate"><span class="pre">SinglePrebuiltReference</span></code></a>) – A pre-built reference created with
<a class="reference internal" href="#singler.build_single_reference.build_single_reference" title="singler.build_single_reference.build_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a>.</p></li>
<li><p><strong>assay_type</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]) – Assay containing the expression matrix,
if <cite>test_data</cite> is a
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>.</p></li>
<li><p><strong>check_missing</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Whether to check for and remove rows with missing (NaN) values
from <code class="docutils literal notranslate"><span class="pre">test_data</span></code>.</p></li>
<li><p><strong>quantile</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#float" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>) – Quantile of the correlation distribution for computing the score for each label.
Larger values increase sensitivity of matches at the expense of
similarity to the average behavior of each label.</p></li>
<li><p><strong>use_fine_tune</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Whether fine-tuning should be performed. This improves accuracy for distinguishing
between similar labels but requires more computational work.</p></li>
<li><p><strong>fine_tune_threshold</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#float" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></a>) – Maximum difference from the maximum correlation to use in fine-tuning.
All labels above this threshold are used for another round of fine-tuning.</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Number of threads to use during classification.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://biocpy.github.io/BiocFrame/api/biocframe.html#biocframe.BiocFrame.BiocFrame" title="(in BiocFrame v0.3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BiocFrame</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A data frame containing the <code class="docutils literal notranslate"><span class="pre">best</span></code> label, the <code class="docutils literal notranslate"><span class="pre">scores</span></code>
for each label (as a nested BiocFrame), and the <code class="docutils literal notranslate"><span class="pre">delta</span></code> from the best
to the second-best label.  Each row corresponds to a column of <code class="docutils literal notranslate"><span class="pre">test</span></code>.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler.fetch_reference">
<span id="singler-fetch-reference-module"></span><h2>singler.fetch_reference module<a class="headerlink" href="#module-singler.fetch_reference" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="singler.fetch_reference.fetch_github_reference">
<span class="sig-prename descclassname"><span class="pre">singler.fetch_reference.</span></span><span class="sig-name descname"><span class="pre">fetch_github_reference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multiple_ids</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/fetch_reference.html#fetch_github_reference"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.fetch_reference.fetch_github_reference" title="Link to this definition">#</a></dt>
<dd><p>Fetch a reference dataset from the
<a class="reference external" href="https://github.com/kanaverse/singlepp-references">pre-compiled GitHub registry</a>,
for use in annotation with other <strong>singler</strong> functions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Literal" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[<code class="docutils literal notranslate"><span class="pre">'BlueprintEncode'</span></code>, <code class="docutils literal notranslate"><span class="pre">'DatabaseImmuneCellExpression'</span></code>, <code class="docutils literal notranslate"><span class="pre">'HumanPrimaryCellAtlas'</span></code>, <code class="docutils literal notranslate"><span class="pre">'MonacoImmune'</span></code>, <code class="docutils literal notranslate"><span class="pre">'NovershternHematopoietic'</span></code>, <code class="docutils literal notranslate"><span class="pre">'ImmGen'</span></code>, <code class="docutils literal notranslate"><span class="pre">'MouseRNAseq'</span></code>]) – Name of the reference dataset.</p></li>
<li><p><strong>cache_dir</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]) – Path to a cache directory in which to store
the files downloaded from the remote. If the files are already
present, the download is skipped.</p></li>
<li><p><strong>multiple_ids</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Whether to report multiple feature IDs.
If True, each feature is represented by a list with zero,
one or more feature identifiers (e.g., for ambiguous mappings).
If False, each feature is represented by a string or None.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The reference dataset as a SummarizedExperiment,
parts of which can be passed to <a class="reference internal" href="#singler.build_single_reference.build_single_reference" title="singler.build_single_reference.build_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a>.</p>
</dd>
</dl>
<p>Specifically, the <code class="docutils literal notranslate"><span class="pre">ranks</span></code> assay of the output can be used as <code class="docutils literal notranslate"><span class="pre">ref</span></code> in
<a class="reference internal" href="#singler.build_single_reference.build_single_reference" title="singler.build_single_reference.build_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a>;
one of the labels in the column data can be used as <code class="docutils literal notranslate"><span class="pre">labels</span></code>;
and one of the gene types in the row data can be used as <code class="docutils literal notranslate"><span class="pre">features</span></code>.</p>
<p>As the ranks are not log-normalized values, users should also use
the relevant pre-computed marker list in the metadata. The selected
marker list should match up with the chosen set of <code class="docutils literal notranslate"><span class="pre">labels</span></code>. In
addition, the markers are stored as row indices and need to be converted
to feature identifiers; this is achieved by passing the marker list to
<a class="reference internal" href="#singler.fetch_reference.realize_github_markers" title="singler.fetch_reference.realize_github_markers"><code class="xref py py-meth docutils literal notranslate"><span class="pre">realize_github_markers()</span></code></a> with the same
gene types that were used in <code class="docutils literal notranslate"><span class="pre">features</span></code>. The output can then be passed
as <code class="docutils literal notranslate"><span class="pre">markers</span></code> in the <cite>build_reference()</cite> call.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">multiple_ids</span> <span class="pre">=</span> <span class="pre">True</span></code>, each <code class="docutils literal notranslate"><span class="pre">row_data</span></code> column will be a list of
lists of possible identifiers for each feature. Callers are responsible for
resolving this list of lists into a list of single identifiers for each
feature, before passing it onto other functions like
<a class="reference internal" href="#singler.build_single_reference.build_single_reference" title="singler.build_single_reference.build_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a>.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="singler.fetch_reference.realize_github_markers">
<span class="sig-prename descclassname"><span class="pre">singler.fetch_reference.</span></span><span class="sig-name descname"><span class="pre">realize_github_markers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">markers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_markers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">restrict_to</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/fetch_reference.html#realize_github_markers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.fetch_reference.realize_github_markers" title="Link to this definition">#</a></dt>
<dd><p>Convert marker indices from a GitHub reference dataset into feature identifiers.  This allows the markers to be
used in <a class="reference internal" href="#singler.build_single_reference.build_single_reference" title="singler.build_single_reference.build_single_reference"><code class="xref py py-meth docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>markers</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]]) – Upregulated markers for each pairwise comparison between labels.
Specifically, <code class="docutils literal notranslate"><span class="pre">markers[a][b]</span></code> should be a sequence of features
that are upregulated in <code class="docutils literal notranslate"><span class="pre">a</span></code> compared to <code class="docutils literal notranslate"><span class="pre">b</span></code>. Features are
represented as indices into <code class="docutils literal notranslate"><span class="pre">features</span></code>.</p></li>
<li><p><strong>features</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>) – Sequence of identifiers for each feature. Features with no valid
identifier for a particular gene type (e.g., no known symbol)
should be represented by None.</p></li>
<li><p><strong>num_markers</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]) – Number of markers to retain. If None, all markers are retained.</p></li>
<li><p><strong>restrict_to</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#set" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">set</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/constants.html#None" title="(in Python v3.9)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]) – Subset of available features to restrict the marker selection.
Only features in <code class="docutils literal notranslate"><span class="pre">restrict_to</span></code> will be reported in the output.
If None, no restriction is performed.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary with the same structure
as <code class="docutils literal notranslate"><span class="pre">markers</span></code>, where each inner sequence contains the corresponding
feature identifiers in <code class="docutils literal notranslate"><span class="pre">features</span></code>. Feature identifiers are guaranteed
to be non-None and to be in <code class="docutils literal notranslate"><span class="pre">restrict_to</span></code> (if specified). Each
inner sequence should have length <code class="docutils literal notranslate"><span class="pre">num_markers</span></code> (or less, if not
enough non-None/restricted identifiers are available).</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler.get_classic_markers">
<span id="singler-get-classic-markers-module"></span><h2>singler.get_classic_markers module<a class="headerlink" href="#module-singler.get_classic_markers" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="singler.get_classic_markers.get_classic_markers">
<span class="sig-prename descclassname"><span class="pre">singler.get_classic_markers.</span></span><span class="sig-name descname"><span class="pre">get_classic_markers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assay_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'logcounts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">restrict_to</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_de</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/get_classic_markers.html#get_classic_markers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.get_classic_markers.get_classic_markers" title="Link to this definition">#</a></dt>
<dd><p>Compute markers from a reference using the classic SingleR algorithm. This is typically done for reference
datasets derived from replicated bulk transcriptomic experiments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_data</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#list" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>]]) – A matrix-like object containing the log-normalized expression values of a reference dataset.
Each column is a sample and each row is a feature.
Alternatively, this can be a <a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a>
containing a matrix-like object in one of its assays.
Alternatively, a list of such matrices or <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code> objects,
typically for multiple batches of the same reference;
it is assumed that different batches exhibit at least some overlap in their <code class="docutils literal notranslate"><span class="pre">features</span></code> and <code class="docutils literal notranslate"><span class="pre">labels</span></code>.</p></li>
<li><p><strong>ref_labels</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#list" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]]) – A sequence of length equal to the number of columns of <code class="docutils literal notranslate"><span class="pre">ref</span></code>,
containing a label (usually a string) for each column.
Alternatively, a list of such sequences of length equal to that of a list <code class="docutils literal notranslate"><span class="pre">ref</span></code>;
each sequence should have length equal to the number of columns of the corresponding entry of <code class="docutils literal notranslate"><span class="pre">ref</span></code>.</p></li>
<li><p><strong>ref_features</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#list" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>]]) – A sequence of length equal to the number of rows of <code class="docutils literal notranslate"><span class="pre">ref</span></code>,
containing the feature name (usually a string) for each row.
Alternatively, a list of such sequences of length equal to that of a list <code class="docutils literal notranslate"><span class="pre">ref</span></code>;
each sequence should have length equal to the number of rows of the corresponding entry of <code class="docutils literal notranslate"><span class="pre">ref</span></code>.</p></li>
<li><p><strong>assay_type</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]) – Name or index of the assay containing the assay of interest,
if <code class="docutils literal notranslate"><span class="pre">ref</span></code> is or contains
<a class="reference external" href="https://biocpy.github.io/SummarizedExperiment/api/summarizedexperiment.html#summarizedexperiment.SummarizedExperiment.SummarizedExperiment" title="(in SummarizedExperiment v0.3.3)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></a> objects.</p></li>
<li><p><strong>check_missing</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>) – Whether to check for and remove rows with missing (NaN) values in the reference matrices.
This can be set to False if it is known that no NaN values exist.</p></li>
<li><p><strong>restrict_to</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#set" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">set</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/constants.html#None" title="(in Python v3.9)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]) – Subset of available features to restrict to. Only features in
<code class="docutils literal notranslate"><span class="pre">restrict_to</span></code> will be used in the reference building. If None,
no restriction is performed.</p></li>
<li><p><strong>num_de</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>]) – Number of differentially expressed genes to use as markers for each pairwise comparison between labels.
If None, an appropriate number of genes is automatically determined.</p></li>
<li><p><strong>num_threads</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Number of threads to use for the calculations.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>[<a class="reference external" href="https://docs.python.org/3.9/library/typing.html#typing.Any" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></a>, <a class="reference external" href="https://docs.python.org/3.9/library/stdtypes.html#list" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code></a>]]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary of dictionary of lists
containing the markers for each pairwise comparison between labels,
i.e., <code class="docutils literal notranslate"><span class="pre">markers[a][b]</span></code> contains the upregulated markers for label
<code class="docutils literal notranslate"><span class="pre">a</span></code> over label <code class="docutils literal notranslate"><span class="pre">b</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="singler.get_classic_markers.number_of_classic_markers">
<span class="sig-prename descclassname"><span class="pre">singler.get_classic_markers.</span></span><span class="sig-name descname"><span class="pre">number_of_classic_markers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_labels</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/singler/get_classic_markers.html#number_of_classic_markers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#singler.get_classic_markers.number_of_classic_markers" title="Link to this definition">#</a></dt>
<dd><p>Compute the number of markers to detect for a given number of labels, using the classic SingleR marker detection
algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_labels</strong> (<a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></a>) – Number of labels.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Number of markers.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3.9/library/functions.html#int" title="(in Python v3.9)">int</a></p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-singler">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-singler" title="Link to this heading">#</a></h2>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          <a class="prev-page" href="modules.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">singler</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Aaron Lun
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">singler package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-singler.annotate_integrated">singler.annotate_integrated module</a><ul>
<li><a class="reference internal" href="#singler.annotate_integrated.annotate_integrated"><code class="docutils literal notranslate"><span class="pre">annotate_integrated()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler.annotate_single">singler.annotate_single module</a><ul>
<li><a class="reference internal" href="#singler.annotate_single.annotate_single"><code class="docutils literal notranslate"><span class="pre">annotate_single()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler.build_integrated_references">singler.build_integrated_references module</a><ul>
<li><a class="reference internal" href="#singler.build_integrated_references.IntegratedReferences"><code class="docutils literal notranslate"><span class="pre">IntegratedReferences</span></code></a><ul>
<li><a class="reference internal" href="#singler.build_integrated_references.IntegratedReferences.reference_labels"><code class="docutils literal notranslate"><span class="pre">IntegratedReferences.reference_labels</span></code></a></li>
<li><a class="reference internal" href="#singler.build_integrated_references.IntegratedReferences.reference_names"><code class="docutils literal notranslate"><span class="pre">IntegratedReferences.reference_names</span></code></a></li>
<li><a class="reference internal" href="#singler.build_integrated_references.IntegratedReferences.test_features"><code class="docutils literal notranslate"><span class="pre">IntegratedReferences.test_features</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#singler.build_integrated_references.build_integrated_references"><code class="docutils literal notranslate"><span class="pre">build_integrated_references()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler.build_single_reference">singler.build_single_reference module</a><ul>
<li><a class="reference internal" href="#singler.build_single_reference.SinglePrebuiltReference"><code class="docutils literal notranslate"><span class="pre">SinglePrebuiltReference</span></code></a><ul>
<li><a class="reference internal" href="#singler.build_single_reference.SinglePrebuiltReference.features"><code class="docutils literal notranslate"><span class="pre">SinglePrebuiltReference.features</span></code></a></li>
<li><a class="reference internal" href="#singler.build_single_reference.SinglePrebuiltReference.labels"><code class="docutils literal notranslate"><span class="pre">SinglePrebuiltReference.labels</span></code></a></li>
<li><a class="reference internal" href="#singler.build_single_reference.SinglePrebuiltReference.marker_subset"><code class="docutils literal notranslate"><span class="pre">SinglePrebuiltReference.marker_subset()</span></code></a></li>
<li><a class="reference internal" href="#singler.build_single_reference.SinglePrebuiltReference.markers"><code class="docutils literal notranslate"><span class="pre">SinglePrebuiltReference.markers</span></code></a></li>
<li><a class="reference internal" href="#singler.build_single_reference.SinglePrebuiltReference.num_labels"><code class="docutils literal notranslate"><span class="pre">SinglePrebuiltReference.num_labels()</span></code></a></li>
<li><a class="reference internal" href="#singler.build_single_reference.SinglePrebuiltReference.num_markers"><code class="docutils literal notranslate"><span class="pre">SinglePrebuiltReference.num_markers()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#singler.build_single_reference.build_single_reference"><code class="docutils literal notranslate"><span class="pre">build_single_reference()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler.classify_integrated_references">singler.classify_integrated_references module</a><ul>
<li><a class="reference internal" href="#singler.classify_integrated_references.classify_integrated_references"><code class="docutils literal notranslate"><span class="pre">classify_integrated_references()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler.classify_single_reference">singler.classify_single_reference module</a><ul>
<li><a class="reference internal" href="#singler.classify_single_reference.classify_single_reference"><code class="docutils literal notranslate"><span class="pre">classify_single_reference()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler.fetch_reference">singler.fetch_reference module</a><ul>
<li><a class="reference internal" href="#singler.fetch_reference.fetch_github_reference"><code class="docutils literal notranslate"><span class="pre">fetch_github_reference()</span></code></a></li>
<li><a class="reference internal" href="#singler.fetch_reference.realize_github_markers"><code class="docutils literal notranslate"><span class="pre">realize_github_markers()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler.get_classic_markers">singler.get_classic_markers module</a><ul>
<li><a class="reference internal" href="#singler.get_classic_markers.get_classic_markers"><code class="docutils literal notranslate"><span class="pre">get_classic_markers()</span></code></a></li>
<li><a class="reference internal" href="#singler.get_classic_markers.number_of_classic_markers"><code class="docutils literal notranslate"><span class="pre">number_of_classic_markers()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-singler">Module contents</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=3f1b9271"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=32e29ea5"></script>
    </body>
</html>